client/kubernetes/src/main/java/org/jolokia/kubernetes/client/KubernetesJmxConnector.java

```diff
@@ -0,0 +1,229 @@
+  }
+
+  /**
+   * @return a connection if successful
+   */
+  protected J4pClient expandAndProbeUrl(ApiClient client,
+      Map<String, Object> env) throws MalformedURLException {
+    Configuration.setDefaultApiClient(client);
+
+    CoreV1Api api = new CoreV1Api();
+    String proxyPath = this.serviceUrl.getURLPath();
+    final HashMap<String, String> headersForProbe = createHeadersForProbe(env);
+    try {
+      if (SERVICE_PATTERN.matcher(proxyPath).matches()) {
+        final Matcher matcher = SERVICE_PATTERN.matcher(proxyPath);
+        if (matcher.find()) {
+          String namespacePattern = matcher.group(1);
+          String servicePattern = matcher.group();
+          String actualNamespace = null;
+          String actualName = null;
+          for (final V1Service service : api
+              .listServiceForAllNamespaces(null, null, false, null, null, null, null, 5, null)
+              .getItems()) {
+
+            if (service.getMetadata().getNamespace().matches(namespacePattern) && service
+                .getMetadata().getName().matches(servicePattern)) {
+              actualNamespace = service.getMetadata().getNamespace();
+              actualName = service.getMetadata().getName();
+            }
+          }
+          if (actualName == null || actualNamespace == null) {
+            throw new MalformedURLException(
+                "Coult not find service in cluster for pattern " + proxyPath);
+          }
+          if (!actualNamespace.equals(namespacePattern)) {
+            proxyPath = proxyPath.replace(namespacePattern, actualNamespace);
+          }
+          if (!actualName.equals(servicePattern)) {
+            proxyPath = proxyPath.replace(servicePattern, actualName);
+          }
+          //probe a request to this URL via proxy
+          try {
+            final Response response = probeProxyPath(client, proxyPath,
+                headersForProbe);
+            response.body().close();
+            if (response.isSuccessful()) {
+              return new J4pClient(
+                  proxyPath, new MinimalHttpClientAdapter(client, proxyPath, env));
+            }
+          } catch (IOException ignore) {
+          }
+          //try to fall back to finding a pod to see if we are more successful connecting to it, will be picked up from next if block if successful
+
+          String path = matcher.group(3);
+          proxyPath = findPodPathIfAnyForService(actualName, actualNamespace, path, api);
+
+
+        }
+      }
+      if (POD_PATTERN.matcher(proxyPath).matches()) {
+        final Matcher matcher = POD_PATTERN.matcher(proxyPath);
+        if (matcher.find()) {
+          String namespacePattern = matcher.group(1);
+
+          String podPattern = matcher.group(2);
+          String actualNamespace = null;
+          String actualPodName = null;
+          for (final V1Pod pod : api
+              .listPodForAllNamespaces(null, null, false, null, null, null, null, 5, null)
```

Comment:
There are 2 usecases.
1. You want to connect to a specific pod for investigation. Use the exact pod name. 
2. You want to connect to any given pod of a service, as they typically display the same behaviour. One pod will be selected at connect time. When disconnected it will naturally stop working, but it depends a bit on the tool behaves, whether stats are not updated or you get explicit errors. A reconnect will connect to the first matching pod.

Reply:
> 2\. as they typically display the same behaviour

Are you sure ? As pods are spread over different nodes and since metric information is stateful, I don't regard the pods as interchangeable wrt/ to metrics. Consider os information from MBean, MBeans that access file systems etf. In a multi-arch cluster the different pods can even run on different architectures (arm vs i86 etc)

Wouldn't a user be surprised that the memory graph make a jump in jconsole if a reconnection happens ? There is no indication by JMX that it is now connected to *another* JVM.

However, I see your use case in favor of convenience. One idea could be that only when going over a service or deployment the JMX connector URL that kind of roulette applies. That way a user has to know what she does. However when using the 'pod' variant of the URL, then a single pod has to be selected. (that would require to also allow a "deployment", "replicaset" (or "rs"), "statefulset" and "job" kind of JMX connectors (but happy to start with service and deployment as they are probably the most common).

wdyt ?

Reply:
I think the risks of using wildcards in pod url can be documented. In my current project the last 2 years shows very homogenous behavior over pods and setup is identical, so personally I would like the convenience of a URL that you can store across deploys. For instance if I inspect a pod or port forward, I tend to pick the top of the list.  For JMC specifically I am thinking of both wizard / browser to interactively connect to pods, and alternatively probe pods in cluster for Jolokia (if user has configured it to). I could look at the possibility of interactively or always hard wire to a given pod.